crates/sui-graphql-rpc/src/types/transaction_block.rs

```diff
@@ -297,102 +305,212 @@ impl TransactionBlock {
-                        );
-
-                        if let Some(a) = &filter.sign_address {
-                            let sub_query = tx_senders::dsl::tx_senders
-                                .select(tx_senders::dsl::tx_sequence_number)
-                                .filter(tx_senders::dsl::sender.eq(a.into_vec()));
-                            query = query.filter(tx::dsl::tx_sequence_number.eq_any(sub_query));
-                        }
-
-                        if let Some(a) = &filter.recv_address {
-                            let sub_query = tx_recipients::dsl::tx_recipients
-                                .select(tx_recipients::dsl::tx_sequence_number)
-                                .filter(tx_recipients::dsl::recipient.eq(a.into_vec()));
-                            query = query.filter(tx::dsl::tx_sequence_number.eq_any(sub_query));
-                        }
-
-                        if let Some(o) = &filter.input_object {
-                            let sub_query = tx_input_objects::dsl::tx_input_objects
-                                .select(tx_input_objects::dsl::tx_sequence_number)
-                                .filter(tx_input_objects::dsl::object_id.eq(o.into_vec()));
-                            query = query.filter(tx::dsl::tx_sequence_number.eq_any(sub_query));
-                        }
+        let db: &Db = ctx.data_unchecked();
+
+        // Anchor the lower and upper checkpoint bounds if provided from the filter, otherwise
+        // default the lower bound to 0 and the upper bound to `checkpoint_viewed_at`. Increment
+        // `after` by 1 so that we can uniformly select the `min_tx_sequence_number` for the lower
+        // bound. Similarly, decrement `before` by 1 so that we can uniformly select the
+        // `max_tx_sequence_number`. In other words, the range consists of all transactions from the
+        // smallest tx_sequence_number in lo_cp to the max tx_sequence_number in hi_cp.
+        let mut lo_cp = max_option!(
+            filter.after_checkpoint.map(|x| x.saturating_add(1)),
+            filter.at_checkpoint,
+            Some(0)
+        )
+        .unwrap(); // Safe to unwrap because of the `Some(0)` default
+
+        let mut hi_cp = min_option!(
+            filter.at_checkpoint,
+            filter.before_checkpoint.map(|x| x.saturating_sub(1)),
+            Some(checkpoint_viewed_at)
+        )
+        .unwrap();
+
+        // If `scan_limit` is set, we need to adjust the lower and upper bounds. It is up to
+        // the caller of `TransactionBlock::paginate` to determine whether `scan_limit` is
+        // required.
+        if let Some(scan_limit) = scan_limit {
+            if page.is_from_front() {
+                hi_cp = std::cmp::min(hi_cp, lo_cp.saturating_add(scan_limit));
+            } else {
+                // Similar to the `first` case, `checkpoint_viewed_at` offers a fixed point to
```

Comment:
Not sure I follow this comment. What's the `first` case? And `checkpoint_viewed_at` is not mentioned on line 357?